### PR TITLE
Fix support for allow_zero_cpus flag

### PR DESCRIPTION
**Pull request by alexis-cousein**
#### Describe Bug or Feature

There is a missing default for allow_zero_cpus in the defaults, and this will make the hook crash when faced with -lselect=1:ncpus=0 jobs unless the cgroup hook config file contains a value for it.

Also, the code brought in was not actually aware of the new execjob_resize, so the code now takes the opportunaity to make sure all corner cases are "clean" in a way that minimises the differences you could see between a system where the cgroup hook is enabled and one where it is not.

#### Describe Your Change

1. For execjob_begin/resize, if zero cpus are not allowed and the cpuset controller is enabled, when no ncpus are requested on mother superior the execjob_begin will reject the job (and execjob_resize will reject the resize). When no cpus are requested on sister MoMs, then the per-job cpuset will be destroyed (which entails an attempt to get rid of processes to correctly support execjob_resize).

2. execjob_attach and execjob_launch will, if the per-job cpuset cgroup is missing (see 1.), either place processes in the main pbspro.slice cpuset if zero cpus are allowed, or simply reject the event if not (with a clear message and a log message). This is done in add_pids, the routine that actually adds the required PIDs to the "correct" cpuset tasks file.

As a result, if zero cpus is NOT allowed and cpusets are enabled, you can still have exec_vnode containing chunks without cpus on some nodes (e.g. to reserve some custom vnode-level resources on them). But you will not be able to launch processes on those nodes.

#### Link to Design Doc

https://pbspro.atlassian.net/wiki/spaces/PD/pages/1627750401/Support+for+the+cpu+cgroup+controller+and+zero-CPU+jobs

Has been revised.

#### Attach Test and Valgrind Logs/Output

This was written to make the cgroup hook pass one of the existing PTL tests which fails:

 test_cgroup_assign_resources_mem_only_vnode

which generates this exception:
04/03/2020 12:22:30.411561;0080;pbs_python;Hook;pbs_python;['Traceback (most recent call last):', ' File "<embedded code object>", line 5479, in main', ' File "<embedded code object>", line 935, in invoke_handler', ' File "<embedded code object>", line 965, in _execjob_begin_handler', ' File "<embedded code object>", line 4641, in configure_job', "KeyError: 'allow_zero_cpus'"]

### CODE REVIEW

File: src/hooks/cgroups/pbs_cgroups.PY [hunk 1]
```diff
@@ -960,9 +960,28 @@ class HookUtils(object):
...
+                               'mother superior: rejecting job')
+                    event.reject('cpuset enabled with mandatory ncpus >= 1, '
+                                 'but job does not request ncpus on '
+                                 'mother superior: rejecting job')
+                else:
+                    pbs.logmsg(pbs.EVENT_JOB_USAGE,
+                               'cpuset enabled with mandatory ncpus >= 1, '
+                               'but job does not request ncpus on host: '
+                               'deleting cpuset cgroup')
+                # will be done in configure_job
```
Comment (mike0042) [inline src/hooks/cgroups/pbs_cgroups.PY:980]: Comment should appear above pbs.logmsg() call.

File: src/hooks/cgroups/pbs_cgroups.PY [hunk 2]
```diff
@@ -1180,6 +1199,24 @@ class HookUtils(object):
...
+                               'mother superior: rejecting resize')
+                    event.reject('cpuset enabled with mandatory ncpus >= 1, '
+                                 'but job does not request ncpus on '
+                                 'mother superior: rejecting resize')
+                else:
+                    pbs.logmsg(pbs.EVENT_ERROR,
+                               'cpuset enabled with mandatory ncpus >= 1, '
+                               'but job no longer requests ncpus on host: '
+                               'deleting cpuset cgroup')
+                # will be done in configure_job
```
Comment (mike0042) [inline src/hooks/cgroups/pbs_cgroups.PY:1219]: Comment should appear above pbs.logmsg() call.

File: src/hooks/cgroups/pbs_cgroups.PY [hunk 3]
```diff
@@ -3444,8 +3482,17 @@ class CgroupUtils(object):
...
+                if (self.cfg['cgroup']['cpuset']['enabled']
+                        and not self.cfg['cgroup']['cpuset']
+                                        ['allow_zero_cpus']):
+                    pbs.logmsg(pbs.EVENT_JOB_USAGE,
+                               'No cpuset cgroup present '
+                               'and npcus==0 disallowed')
+                    pbs.event().reject('No cpuset cgroup present '
+                                       'and npcus==0 disallowed')
+                else:
+                    # zero CPU  job: should be attached to root pbspro cpuset
```
Comment (mike0042) [inline src/hooks/cgroups/pbs_cgroups.PY:3494]: Two spaces after CPU

**Comment by lisa-altair** [comment]
This is the before and after test results:
[before.txt](https://github.com/PBSPro/pbspro/files/4525970/before.txt)
[after_fix.txt](https://github.com/PBSPro/pbspro/files/4525969/after_fix.txt)

File: src/hooks/cgroups/pbs_cgroups.PY (hunk 1 above)
Comment (alexis-cousein) [reply src/hooks/cgroups/pbs_cgroups.PY]: Corrected.

File: src/hooks/cgroups/pbs_cgroups.PY (hunk 2 above)
Comment (alexis-cousein) [reply src/hooks/cgroups/pbs_cgroups.PY]: Corrected.

File: src/hooks/cgroups/pbs_cgroups.PY (hunk 3 above)
Comment (alexis-cousein) [reply src/hooks/cgroups/pbs_cgroups.PY]: Corrected (one space only)

**Comment by alexis-cousein** [comment]
Changes requested by @mike0042 implemented.

**Pull request merged**